dm/worker/server.go

```diff
@@ -464,3 +468,69 @@ func (s *Server) splitHostPort() (host, port string, err error) {
 	}
 	return
 }
+
+// unifyMasterBinlogPos eliminates different masterBinlog in one response
+// see https://github.com/pingcap/dm/issues/727
+func unifyMasterBinlogPos(resp *pb.QueryStatusResponse) {
+	var (
+		syncStatus          []*pb.SubTaskStatus_Sync
+		syncMasterBinlog    []*mysql.Position
+		lastestMasterBinlog mysql.Position // not pointer, to make use of zero value and avoid nil check
+		relayMasterBinlog   *mysql.Position
+	)
+
+	// uninitialized mysql.Position is less than any initialized mysql.Position
+	if resp.RelayStatus.Stage != pb.Stage_Stopped {
+		var err error
+		relayMasterBinlog, err = utils.DecodeBinlogPosition(resp.RelayStatus.MasterBinlog)
+		if err != nil {
+			log.L().Error("failed to decode relay's master binlog position", zap.Stringer("response", resp), zap.Error(err))
+			return
+		}
+		lastestMasterBinlog = *relayMasterBinlog
+	}
+
+	for _, stStatus := range resp.SubTaskStatus {
+		if stStatus.Unit == pb.UnitType_Sync {
+			s := stStatus.Status.(*pb.SubTaskStatus_Sync)
+			syncStatus = append(syncStatus, s)
+
+			position, err := utils.DecodeBinlogPosition(s.Sync.MasterBinlog)
+			if err != nil {
+				log.L().Error("failed to decode sync's master binlog position", zap.Stringer("response", resp), zap.Error(err))
+				return
+			}
+			if lastestMasterBinlog.Compare(*position) < 0 {
+				lastestMasterBinlog = *position
+			}
+			syncMasterBinlog = append(syncMasterBinlog, position)
+		}
+	}
+
+	// re-check relay
+	if resp.RelayStatus.Stage != pb.Stage_Stopped && lastestMasterBinlog.Compare(*relayMasterBinlog) != 0 {
+		relayPos, err := utils.DecodeBinlogPosition(resp.RelayStatus.RelayBinlog)
+		if err != nil {
+			log.L().Error("failed to decode relay binlog position", zap.Stringer("response", resp), zap.Error(err))
+			return
+		}
+		catchUp := lastestMasterBinlog.Compare(*relayPos) == 0
```

Comment:
If `EnableGTID == true`, we compare GTID to check `catchUp` because in GTID mode, some unimportant binlog events may be delayed before being sent to DM from MySQL. 😢 see https://github.com/pingcap/dm/blob/019f4995eec7a17cd95ee1dddf1b113b0436675f/relay/relay.go#L727-L733

Reply:
addressed in https://github.com/pingcap/dm/pull/817/commits/2b09abe4daa0e85ccdf39a569a1e8bae80e0ad75